### PR TITLE
Expose recompute configs to the frontend

### PR DESCRIPTION
**Pull request by SherlockNoMad**
**Description**: Expose recompute configs to the frontend

### CODE REVIEW

File: orttraining/orttraining/python/training/orttrainer_options.py [hunk 1]
```diff
@@ -116,6 +116,30 @@ class ORTTrainerOptions(object):
...
+                        },
+                        'transformer_layer': {
+                            'type': 'boolean',
+                            'default': False
+                        }, 
+                        'number_layers': {
+                            'type': 'integer',
+                            'min': 0,
+                            'default': 0
+                        }
```
Comment (KeDengMS) [inline orttraining/orttraining/python/training/orttrainer_options.py:140]: Can these two options merge into num_transformer_layers? So recompute.num_transformer_layers ==0 means no recompute of layers. Besides, can we use both dropout/gelu recompute and layer recompute?

**Comment by KeDengMS** [comment]
Please add some FE tests in orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py for the new options.

**Review by KeDengMS — DISMISSED** [review]
:shipit:

**Review by KeDengMS — DISMISSED** [review]
:shipit:

File: orttraining/orttraining/python/training/orttrainer_options.py [hunk 2]
```diff
@@ -435,6 +469,30 @@ def _check_is_callable(field, value, error):
             }
         }
     },
+    'recompute': {
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer_options.py:472]: Is this the most intuitive category? SHould it be under a generic "graph" category, maybe?

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 3]
```diff
@@ -728,6 +734,61 @@ def testORTTrainerMixedPrecisionLossScaler(seed, device, expected_loss, fetches)
     assert trainer._onnx_model is not None
 
 
+def recompute_test_data():
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:737]: done

File: orttraining/orttraining/python/training/orttrainer_options.py (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/orttrainer_options.py:472]: good idea.

File: orttraining/orttraining/python/training/orttrainer_options.py [hunk 4]
```diff
@@ -221,6 +245,16 @@ class ORTTrainerOptions(object):
...
+        recompute (dict):
+            recompute options 
+        recompute.attn_dropout (bool, default is False):
+            enable attention dropout recompute
+        recompute.gelu (bool, default is False):
+            enable gelu recompute
+        recompute.transformer_layer (bool, default is False):
+            enable transformer layerwise recompute 
+        recompute.number_layers (int, default is 0)
+            number of layers to apply recompute
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/orttrainer_options.py]: done

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 5]
```diff
@@ -94,6 +94,12 @@ def testORTTrainerOptionsDefaultValues(test_input):
             'enabled': False,
             'loss_scaler': None
         },
+        'recompute': {
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:97]:
```suggestion
        'graph_transformer': {
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 6]
```diff
@@ -728,6 +734,61 @@ def testORTTrainerMixedPrecisionLossScaler(seed, device, expected_loss, fetches)
...
+    device = 'cuda'
+    rtol = 1e-3
+    total_steps = len(expected_loss)
+    torch.manual_seed(seed)
+    set_seed(seed)
+
+    # Setup ORTTrainer
+    loss_scaler = amp.DynamicLossScaler()
+    options = orttrainer.ORTTrainerOptions({'device' : {'id' : device},
+                                            'recompute' : {
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:769]:
```suggestion
                                            'graph_transformer' : {
```

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 7]
```diff
@@ -728,6 +734,61 @@ def testORTTrainerMixedPrecisionLossScaler(seed, device, expected_loss, fetches)
...
+def testORTTrainerRecompute(attn_dropout, gelu, transformer_layer, number_layers, expected_loss):
+    seed = 321
+    device = 'cuda'
+    rtol = 1e-3
+    total_steps = len(expected_loss)
+    torch.manual_seed(seed)
+    set_seed(seed)
+
+    # Setup ORTTrainer
+    loss_scaler = amp.DynamicLossScaler()
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:767]:
```suggestion
```
This line is dead code. Either enabled mixed precision or dont use loss scaler. `amp.DynamicLossScaler()` without custom arguments dont need to be instantiated - `{'mixed_precision' : True}` does it by default.

Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:767]: will fix in my next PR.

**Review by thiagocrepaldi — APPROVED** [review]
As synced offline, a dead code line will be removed in a next PR that is coming soon

**Pull request merged**